dropshot/src/server.rs

```diff
@@ -149,15 +201,208 @@ impl<C: ServerContext> HttpServerStarter<C> {
         let local_addr = server.local_addr();
         info!(app_state.log, "listening"; "local_addr" => %local_addr);
 
-        Ok(HttpServerStarter {
+        Ok(InnerHttpServerStarter {
             app_state,
             server,
             local_addr,
         })
     }
+}
 
-    pub fn app_private(&self) -> &C {
-        &self.app_state.private
+struct TlsConn {
+    stream: TlsStream<TcpStream>,
+    remote_addr: SocketAddr,
+}
+
+impl TlsConn {
+    fn new(stream: TlsStream<TcpStream>, remote_addr: SocketAddr) -> TlsConn {
+        TlsConn {
+            stream,
+            remote_addr,
+        }
+    }
+
+    fn remote_addr(&self) -> SocketAddr {
+        self.remote_addr
+    }
+}
+
+impl tokio::io::AsyncRead for TlsConn {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        ctx: &mut core::task::Context,
+        buf: &mut ReadBuf,
+    ) -> Poll<std::io::Result<()>> {
+        let pinned = Pin::new(&mut self.stream);
+        pinned.poll_read(ctx, buf)
+    }
+}
+
+impl tokio::io::AsyncWrite for TlsConn {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        ctx: &mut core::task::Context,
+        data: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let pinned = Pin::new(&mut self.stream);
+        pinned.poll_write(ctx, data)
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        ctx: &mut core::task::Context,
+    ) -> Poll<std::io::Result<()>> {
+        let pinned = Pin::new(&mut self.stream);
+        pinned.poll_flush(ctx)
+    }
+
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        ctx: &mut core::task::Context,
+    ) -> Poll<std::io::Result<()>> {
+        let pinned = Pin::new(&mut self.stream);
+        pinned.poll_shutdown(ctx)
+    }
+}
+
+struct HttpsAcceptor {
+    acceptor: Box<dyn Stream<Item = std::io::Result<TlsConn>> + Send + Unpin>,
+}
+
+impl hyper::server::accept::Accept for HttpsAcceptor {
```

Comment:
I noticed that `Accept` is also implemented for [AddrIncoming](https://docs.rs/hyper/0.14.11/hyper/server/conn/struct.AddrIncoming.html) - is there something which prevents us from using that structure directly?

Reply:
(more on this thread down below)